src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,869 @@
+
+  for(auto it : m)
+  {
+    // We should group constants together...
+    const exprt &t=it.first;
+    int second=negated?(-it.second):it.second;
+    if(t.id()==ID_constant)
+    {
+      std::string value(to_constant_expr(t).get_value().c_str());
+      constants+=binary2integer(value, true)*second;
+    }
+    else
+    {
+      if(second!=0)
+      {
+        if(second==-1)
+        {
+          if(sum==refined_string_typet::index_of_int(0))
+            sum=unary_minus_exprt(t);
+          else
+            sum=minus_exprt(sum, t);
+        }
+        else if(second==1)
+        {
+          if(sum==refined_string_typet::index_of_int(0))
+            sum=t;
+          else
+            sum=plus_exprt(sum, t);
+        }
+      }
+      else
+      {
+        debug() << "in string_refinementt::sum_of_map:"
+                << " warning: several occurences of the same variable: "
+                << t.pretty() << eom;
+        variable_with_multiple_occurence_in_index=true;
+        if(second>1)
+        {
+          for(int i=0; i<second; i++)
+            sum=plus_exprt(sum, t);
+        }
+        else
+        {
+          for(int i=0; i>second; i--)
+            sum=minus_exprt(sum, t);
+        }
+      }
+    }
+  }
+
+  exprt index_const=from_integer(constants, refined_string_typet::index_type());
+  return plus_exprt(sum, index_const);
+}
+
+exprt string_refinementt::simplify_sum(const exprt &f)
+{
+  std::map<exprt, int> map=map_representation_of_sum(f);
+  return sum_over_map(map);
+}
+
+exprt string_refinementt::compute_subst
+(const exprt &qvar, const exprt &val, const exprt &f)
+{
+  exprt positive, negative;
+  // number of time the element should be added (can be negative)
+  // qvar has to be equal to val - f(0) if it appears positively in f
+  // (ie if f(qvar)=f(0) + qvar) and f(0) - val if it appears negatively
+  // in f. So we start by computing val - f(0).
+  std::map<exprt, int> elems=map_representation_of_sum(minus_exprt(val, f));
+
+  // true if qvar appears negatively in f (positively in elems):
+  bool neg=false;
+
+  auto it=elems.find(qvar);
+  assert(it!=elems.end());
+  if(it->second==1 || it->second==-1)
+  {
+    neg=(it->second==1);
+  }
+  else
+  {
+    assert(it->second==0);
+    debug() << "in string_refinementt::compute_subst:"
+            << " warning: occurences of qvar canceled out " << eom;
+  }
+
+  elems.erase(it);
+  return sum_over_map(elems, neg);
+}
+
+
+
+class find_qvar_visitort: public const_expr_visitort
+{
+private:
+  const exprt &qvar_;
+
+public:
+  explicit find_qvar_visitort(const exprt &qvar): qvar_(qvar) {}
+
+  void operator()(const exprt &expr)
+  {
+    if(expr==qvar_)
+      throw true;
+  }
+};
+
+// Look for the given symbol in the index expression
+static bool find_qvar(const exprt index, const symbol_exprt & qvar)
+{
+  find_qvar_visitort v2(qvar);
+  try
+  {
+    index.visit(v2);
+    return false;
+  }
+  catch (bool found) {return found;}
+}
```

Comment:
I would strongly favour an implementation that does not use communication-via-exceptions.